3.0.0 rc erikvk (#270)
Starcounter/Docs: docs/database-classes.md

## Defining database classes

We can define the schema of a Starcounter database table in an application by defining **public abstract C# classes** decorated with the `DatabaseAttribute` attribute. We call these types **database classes**. The names and namespaces we give to these C# classes will be the names and namespaces of the corresponding database tables.

```csharp
using System;
using Starcounter.Database;

[Database]
public abstract class Person
{
    public abstract string Name { get; set; }
}
```

All instances of these database classes created with the `IDatabaseContext.Insert<T>()` method are stored in the Starcounter database persistently.

## Constructors

Database classes support default constructors, which is useful for adding logic that is run each time an instance of the database class is created. Public default constructors of database classes are called whenever a new instance of the class is created with `IDatabaseContext.Insert<T>()`.

For exaαmple, here is our `Person` class with a custom default constructor:

```csharp
using System;
using Starcounter.Database;

[Database]
public abstract class Person
{
    public abstract string Name { get; set; }

>     public abstract DateTime CreatedAt { get; set; }

Reviewer: ```suggestion
    public abstract DateTime CreatedUtcAt { get; set; }
```

Author: Done

    public Person()
    {

>         CreatedAt = DateTime.UtcNow;

Reviewer: ```suggestion
        CreatedUtcAt = DateTime.UtcNow;
```

Author: Done

    }
}
```

> It's allowed to have non-default constructors in database class definitions as well, although they are not called by `IDatabaseContext.Insert<T>()`. Instead they can be useful in unit testing scenarios when you want to inject dependencies or other arguments into a class instance. If you add a constructor with parameters to a database class, you also have to add a default constructor.

Reviewer: I am a bit confused here. What is the point of having such constructors in abstract classes? Would not it make more sense to have the custom constructors in the derived non database classes, defined for the sake of unit testing?

Author: I guess it was a relic from previous versions of the docs. I'll remove that section

Author: Done

## Fields and properties

> Table columns are defined in database classes by abstract instance auto-implemented properties with public get and set accessors and with one of the [supported data types](database-types). The `Person` database class below defines two such columns, `Name` and `CreatedAt`:

Reviewer: ```suggestion
Table columns are defined in database classes by abstract instance auto-implemented properties with public `get` and `set` accessors and with one of the [supported data types](database-types). The `Person` database class below defines two such columns, `Name` and `CreatedUtcAt`:
```

Author: Fixed

```csharp
using System;
using Starcounter.Database;

[Database]
public abstract class Person
{
    public abstract string Name { get; set; }

>     public abstract DateTime CreatedAt { get; set; }

Reviewer: ```suggestion
    public abstract DateTime CreatedUtcAt { get; set; }
```

}
```

> Database classes can also define calculated properties, which are properties that do not expose an instance field. Commonly, we use them to reflect an aspect of the state of a database class instance. Let's say we want a shorthand for calculating the length of the `Name` string of a `Person`. We could do this by introducing a calculated property. These computed properties do not have the same restrictions as column properties, since they are not bound to the corresponding database table.

Reviewer: ```suggestion
Database classes can also define calculated properties, which are properties that do not expose an instance field. Commonly, we use them to reflect an aspect of the state of a database class instance. Let's say we want a shorthand for calculating the length of the `Name` string of a `Person`. We could do this by introducing a calculated property. These computed properties do not have the same restrictions as column properties, since they are not persistent and are not bound to the corresponding database table.
```

```csharp
var ddlExecutor = services.GetRequiredService<IDdlExecutor>();

ddlExecutor.Execute("CREATE INDEX IX_Person_FirstName ON Person (FirstName)");
```

A single property index can also be registered for a column by decorating its associated C# property with the `IndexAttribute` attribute:

```csharp
using System;
using Starcounter.Database;

[Database]
public abstract class Person
{
    [Index]
    public abstract string Name { get; set; }
}
```

**Note**: DDL statements can only be performed from `IDdlExecutor`.

## Relations

### One-to-many relations

It's recommended to model one-to-many relations using references both ways, with the child entity having a reference to the parent and the parent having an instance method or computed property that selects all the children (like in the example with the `Mother`/`Children` relation above).

### Many-to-many relations

Many-to-many relations are best modelled using an **association class**.

> Let's say we want to model many-to-many relation of **share ownership**, as understood as a relation between `Person` and `Company` entities such that a single person can hold multiple shares in multiple companies, and a single company can have multiple shareholders. For this, let's use the association class `HoldsShares`. We can then make queries to this database class to calculate useful information in `Person` and `Company` that is exposed using computed properties. Lastly, we make sure that indexes are registered for the `Owner` and `Equity` properties of `HoldsShares` since we will be quering them a lot.

Reviewer: `Shareholder` should be a better name for this table, instead of `HoldsShares`.

Author: A shareholder is a relation between a person and a company, with quantity? No, the shareholder is the person holding the share, participating as the person in the relation. There is no obvious name for the relation, we could always go with `ShareOwnership` or something.

Reviewer: I would still call it just a `Shareholder` with quantity :), but let's keep it as is.

Author: But it's not a shareholder. It's the relation between a shareholder and the shares he or she holds. You are not a Starcounter employment, you're an employee :)

Reviewer: But MongoDB is web scaled ©.

>It's the relation between a shareholder and the shares he or she holds.

It's a relation between holder & shares. Having such a relation with quantity > 0 makes a holder to be a shareholder :).

Whatever, let's keep it your way, I am fine with it, just thought `Shareholder` is a neat word.

            return db.Sql<Company>("SELECT s.Equity FROM HoldsShares s WHERE s.Owner = ?", this);
        }
    }
}

[Database]
public abstract class Company
{
    public abstract string Name { get; set; }
    public abstract Person Ceo { get; set; }

    public IEnumerable<Person> Shareholders
    {
        get
        {
            var db = DbProxy.GetContext(this);
            return db.Sql<Person>("SELECT s.Owner FROM HoldsShares s WHERE s.Equity = ?", this);
        }
    }
}

[Database]
public abstract class HoldsShares
{
    [Index]
    public abstract Person Owner { get; set; }

    [Index]
    public abstract Company Equity { get; set; }

    public abstract int Quantity { get; set; }
}
```

### Inheritance

Inheritance is supported between database classes and has the same general semantics as class inheritence in C#.

```csharp
using Starcounter.Database;

[Database]
public abstract class Customer
{
   public abstract string Name { get; set; }
}

public abstract class PrivateCustomer : Customer
{
   public abstract string Gender { get; set; }
}

public abstract class CorporateCustomer : Customer
{
   public abstract string VatNumber { get; set; }
}
```

> The `DatabaseAttribute` attribute decoration is inherited from any base class to its subclasses, meaning that any class that directly or indirectly inherits a class that is decorated with `DatabaseAttribute` becomes a database class. In the example above, both `PrivateCustomer` and `CorporateCustomer` become database classes due to them inheriting `Customer`.

Reviewer: ```suggestion
The `DatabaseAttribute` attribute decoration is inherited from any base class to its subclasses, meaning that any class that directly or indirectly inherits a class that is decorated with the `DatabaseAttribute` becomes a database class. In the example above, both `PrivateCustomer` and `CorporateCustomer` become database classes due to them inheriting `Customer`.
```

> This also means that all rows in the `CorporateCustomer` table also are contained in the `Customer` table. The result of the `SELECT C FROM Customer c` SQL query will contain all rows from `Customer` as well as all tables from subclasses of `Customer`.

Reviewer: ```suggestion
This also means that all rows in the `CorporateCustomer` table also are contained in the `Customer` table. The result of the `SELECT C FROM Customer c` SQL query will contain all rows from `Customer` as well as all rows from sub-classes of `Customer`.
```

Author: Thanks for spotting the "tables" typo. Regarding "subclasses", it's "[subclass](https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)#Subclasses_and_superclasses)", not "sub-class"

A database class cannot inherit from a class that's not a database class. It's also not possible to cast a non-database class to a database class.

## Database object identity

Starcounter automatically assigns an unique integer identifier called `Oid` to each database object. The key is unique across all table rows in the entire database. The `IDatabaseContext` type defines methods for working with object `Oid`'s. For the examples above, let's assume that `db` holds a reference to an instance of `IDatabaseContext`, recieved from an `ITransactor` when starting a new database transaction.

### Get the `Oid` from a database object

```csharp
var newObject = db.Insert<Product>();
ulong oid = db.GetOid(newObject);
```

### Find an object in the database by its unique key

```csharp
var oid = 459123UL;
var product = db.Get<Product>(oid);
```

### Notes

- Zero (`0`) is not a valid `Oid`.

> - It's not possible to insert a database object with a predefined `Oid`.

Reviewer: ```suggestion
- It's not yet possible to insert a database object with a predefined `Oid`.
```

Author: Done
